hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -992,12 +1003,34 @@ def add(self, data):
             self.best_status = data.status
             self.reset()
 
-        self.fresh_examples.append(data)
-        if len(self) > self.pool_size:
-            pop_random(self.random, self.used_examples or self.fresh_examples)
-            assert self.pool_size == len(self)
+        for label in data.target_observations:
+            pool = self.scored_examples[label]
+            pool.append(data)
+            pool.sort(key=lambda d: d.target_observations[label], reverse=True)
```

Comment:
Resorting the pool each time we add to it seems suboptimal. Maybe use a heap from the `heapq` module so we maintain the top n largest examples?

Reply:
`timsort` is linear-time for mostly-sorted data, so in practice this is a scan plus a binary search to insert the last element (*if* it's not already in place, which is more common).  The pool is also not that large!

Given that we do want elements ordered by decreasing score and sorting is more than fast enough, I went with clarity over optimisation.